diff --git a/kannel_meta/gw/smsc/smpp_pdu.c b/kannel_meta/gw/smsc/smpp_pdu.c
index 2a46e88..87924b6 100644
--- a/kannel_meta/gw/smsc/smpp_pdu.c
+++ b/kannel_meta/gw/smsc/smpp_pdu.c
@@ -587,6 +587,12 @@ err:
 
 void smpp_pdu_dump(SMPP_PDU *pdu)
 {
+
+	  // No need enquire_link debug info
+    if(pdu->type == 0x80000015) {
+        return;
+    }
+
     debug("sms.smpp", 0, "SMPP PDU %p dump:", (void *) pdu);
     debug("sms.smpp", 0, "  type_name: %s", pdu->type_name);
     switch (pdu->type) {
diff --git a/kannel_meta/gw/smsc/smsc_smpp.c b/kannel_meta/gw/smsc/smsc_smpp.c
index e4a0e73..1fc3f07 100644
--- a/kannel_meta/gw/smsc/smsc_smpp.c
+++ b/kannel_meta/gw/smsc/smsc_smpp.c
@@ -969,7 +969,7 @@ static void send_enquire_link(SMPP *smpp, Connection *conn, long *last_sent)
     *last_sent = date_universal_now();
 
     pdu = smpp_pdu_create(enquire_link, counter_increase(smpp->message_id_counter));
-    dump_pdu("Sending enquire link:", smpp->conn->id, pdu);
+    // dump_pdu("Sending enquire link:", smpp->conn->id, pdu);
     os = smpp_pdu_pack(pdu);
     if (os)
 	conn_write(conn, os); /* Write errors checked by caller. */
@@ -1924,5 +1924,8 @@ static void io_thread(void *arg)
            } else if (ret == 1) { /* data available */
                /* Deal with the PDU we just got */
-                dump_pdu("Got PDU:", smpp->conn->id, pdu);
+                // dump_pdu("Got PDU:", smpp->conn->id, pdu);
+                if(pdu->type != 0x80000015){
+                    dump_pdu("Got PDU:", smpp->conn->id, pdu);
+                }
                ret = handle_pdu(smpp, conn, pdu, &pending_submits);
                smpp_pdu_destroy(pdu);
